libs2eplugins/src/s2e/Plugins/Models/BaseFunctionModels.cpp

```diff
@@ -692,42 +692,67 @@ bool BaseFunctionModels::strcatHelper(S2EExecutionState *state, const uint64_t s
 
     return true;
 }
-
 bool BaseFunctionModels::strstrHelper(S2EExecutionState *state, uint64_t haystackAddr, uint64_t needleAddr, ref<Expr> &retExpr, uint32_t byte_width) {
-    getWarningsStream(state) << "Enter strstr yes\n";
+    getWarningsStream(state) << "Entering strstrHelper\n";
+
     size_t haystackLen, needleLen;
-    if (!findNullCharWithWidth(state, haystackAddr, haystackLen, byte_width) || !findNullCharWithWidth(state, needleAddr, needleLen, byte_width)) {
+    if (!findNullCharWithWidth(state, haystackAddr, haystackLen, byte_width) ||
+        !findNullCharWithWidth(state, needleAddr, needleLen, byte_width)) {
         getWarningsStream(state) << "Failed to find nullptr in haystack or needle\n";
         return false;
     }
 
-    const Expr::Width ret_width = state->getPointerSize() * CHAR_BIT;
     if (needleLen == 0) {
-        retExpr = E_CONST(haystackAddr, ret_width);
-        return true;
+        retExpr = E_CONST(haystackAddr, state->getPointerSize() * CHAR_BIT);
+        return true;  // Immediate return if needle is an empty string
     }
 
-    const ref<Expr> nullExpr = E_CONST(0, ret_width);
+    const Expr::Width pointerWidth = state->getPointerSize() * CHAR_BIT;
+    const ref<Expr> nullExpr = E_CONST(0, pointerWidth);
 
-    ref<Expr> finalExpr = nullExpr; // Start with the assumption that we will not find the needle
-    for (size_t i = 0; i <= haystackLen - needleLen; i += byte_width) {
-        uint64_t strAddrs[2] = {haystackAddr + i, needleAddr};
-        ref<Expr> memcmpResult;
-
-        // Use strncmpHelper to compare the substring of haystack with needle
-        if (!memcmpHelper(state, strAddrs, needleLen, memcmpResult)) {
-            getWarningsStream(state) << "Failed to compare substring of haystack with needle " << needleLen << "\n";
+    std::vector<size_t> badCharSkip(256, needleLen);  // Bad character skip array
+    uint8_t charByte = 0;  // Temporary variable for reading bytes
+    for (size_t i = 0; i < needleLen - 1; ++i) {
+        uint64_t readAddr = needleAddr + i * byte_width;
+        if (!state->mem()->read(readAddr, &charByte, VirtualAddress, true)) {
```

Comment:
For this one, I think I intended to use the first template, but I have not tested if it works as intended. I only observe that there will be mem access

Reply:
But I do think that for simple template functions where the template type is directly tied to the type of one of the function's parameters, you don't need to explicitly specify the template type because the compiler can deduce it.

Reply:
> But I do think that for simple template functions where the template type is directly tied to the type of one of the function's parameters, you don't need to explicitly specify the template type because the compiler can deduce it.

Oh! you are right.